thread/linux/omrthreadnuma.c

```diff
@@ -134,42 +141,100 @@ initializeNumaNodeData(omrthread_library_t threadLibrary, uintptr_t numNodes)
 	} else {
 		DIR *nodes = NULL;
 		unsigned long nodeIndex = 0;
+		const char NODE_PATH[] = "/sys/devices/system/node/";
 		for (nodeIndex = 0; nodeIndex <= numNodes; nodeIndex++) {
 			CPU_ZERO(&numaNodeData[nodeIndex].cpu_set);
 			numaNodeData[nodeIndex].cpu_count = 0;
 		}
 
-		nodes = opendir("/sys/devices/system/node/");
+		nodes = opendir(NODE_PATH);
 		if (NULL == nodes) {
 			result = -1;
 		} else {
 			struct dirent *node = readdir(nodes);
-			while (NULL != node) {
-				if (1 == sscanf(node->d_name, "node%lu", &nodeIndex)) {
-					if (nodeIndex < numNodes) {
-						DIR *cpus = NULL;
-						char nodeName[sizeof("/sys/devices/system/node/") + NAME_MAX + 1] = "/sys/devices/system/node/";
-						strcat(nodeName, node->d_name);
-						cpus = opendir(nodeName);
-						if (NULL != cpus) {
-							struct dirent *cpu = readdir(cpus);
-							while (NULL != cpu) {
-								unsigned long cpuIndex = 0;
-								if (1 == sscanf(cpu->d_name, "cpu%lu", &cpuIndex)) {
-									CPU_SET(cpuIndex, &numaNodeData[nodeIndex + 1].cpu_set);
-									numaNodeData[nodeIndex + 1].cpu_count += 1;
-
-									/* add all cpus to the synthetic 0 node */
-									CPU_SET(cpuIndex, &numaNodeData[0].cpu_set);
-									numaNodeData[0].cpu_count += 1;
+			if (NULL != node) {
+				const char CPUMAP[] = "/cpumap";
+				char pathBuffer[sizeof(NODE_PATH) + sizeof(CPUMAP) + 16]; /* extra space for "node<n>" & terminating null */
+				strcpy(pathBuffer, NODE_PATH);
+				char *nodeRoot = pathBuffer + sizeof(NODE_PATH) - 1; /* shortcut to the end of the string */
+				do {
+					if (1 == sscanf(node->d_name, "node%lu", &nodeIndex)) {
+						if (nodeIndex < numNodes) {
+							strcpy(nodeRoot, node->d_name);
+							strcat(nodeRoot, CPUMAP);
+							int cpumapFile = open(pathBuffer, O_RDONLY);
+							if (-1 != cpumapFile) {
+								char defaultMapBuffer[128];
+								char * mapBuffer = defaultMapBuffer;
+								size_t bufferSize = sizeof(defaultMapBuffer);
+								size_t bytesRead = read(cpumapFile, mapBuffer, bufferSize);
+								if (bytesRead == bufferSize) { /* buffer possibly not big enough */
+									bufferSize = 4096;
+									mapBuffer = malloc(bufferSize);
+									if (NULL == mapBuffer) {
+										result = -1;
+										close(cpumapFile);
+										break;
+									}
+									size_t bytesRead = read(cpumapFile, mapBuffer, bufferSize);
+								}
+								if (0 != bytesRead) {
+									/* scan from the end of the line (lowest numbered CPU) toward the front */
+									char *cursor = mapBuffer+strlen(mapBuffer);
```

Comment:
That would assume that read() would return the string with no trailing crud, e.g. multiple nulls.
I should, however, read bufferSize-1 bytes and add a null at mapBuffer[bytesRead].